contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala

```diff
@@ -0,0 +1,324 @@
+    } finally {
+      input_stream.close()
+    }
+  }
+
+  private def resolvePath(
+      path: Path,
+      hadoopConf: Configuration
+  ): (FileSystem, Path) = {
+    val fs = path.getFileSystem(getHadoopConfiguration)
+    val resolvedPath = stripUserInfo(fs.makeQualified(path))
+    (fs, resolvedPath)
+  }
+
+  private def stripUserInfo(path: Path): Path = {
+    val uri = path.toUri
+    val newUri = new URI(
+      uri.getScheme,
+      null,
+      uri.getHost,
+      uri.getPort,
+      uri.getPath,
+      uri.getQuery,
+      uri.getFragment
+    )
+    new Path(newUri)
+  }
+
+  /**
+   * Method for assuring consistency on filesystem according to the external cache.
+   * Method tries to rewrite TransactionLog entry from temporary path if it does not exists.
+   * Method returns completed [[ExternalCommitEntry]]
+   */
+
+  private def fixDeltaLog(fs: FileSystem, entry: ExternalCommitEntry): Unit = {
+    if (entry.complete) {
+      return
+    }
+    var retry = 0;
+    var copied = false;
+
+    while (true) {
+      logDebug(
+        s"${if (retry > 0) "re" else ""}trying to fix: ${entry.fileName}"
+      )
+      try {
+        if (!copied && !fs.exists(entry.absoluteJsonPath)) {
+          fixDeltaLogCopyTempFile(fs, entry.absoluteTempPath, entry.absoluteJsonPath);
+          copied = true;
+        }
+        fixDeltaLogPutCompleteDbEntry(entry);
+        logInfo(s"fixed ${entry.fileName}")
+        return
+      } catch {
+        case e: Throwable =>
+          logInfo(s"${e.getClass.getName}: $e")
+          if (retry >= 3) throw e
+      }
+      retry += 1;
+    }
+  }
+
+  /**
+   * The following four methods are extracted for testing purposes
+   * so we can more easily inject errors and test for failures.
+   */
+
+  protected def writeCopyTempFile(fs: FileSystem, src: Path, dst: Path) = {
+    copyFile(fs, src, dst)
+  }
+
+  protected def writePutCompleteDbEntry(entry: ExternalCommitEntry): Unit = {
+    putExternalEntry(entry.asComplete(), overwrite = true)
+  }
+
+  protected def fixDeltaLogCopyTempFile(fs: FileSystem, src: Path, dst: Path): Unit = {
+    copyFile(fs, src, dst);
+  }
+
+  protected def fixDeltaLogPutCompleteDbEntry(entry: ExternalCommitEntry): Unit = {
+    putExternalEntry(entry.asComplete(), overwrite = true);
+  }
+
+  /**
+   * Returns path stripped user info.
+   */
+  protected def getPathKey(resolvedPath: Path): Path = {
+    stripUserInfo(resolvedPath)
+  }
+
+  /**
+   * Generate temporary path for TransactionLog.
+   */
+  protected def createTemporaryPath(path: Path): String = {
+    val uuid = java.util.UUID.randomUUID().toString
+    s".tmp/${path.getName}.$uuid"
+  }
+
+  protected def getTablePath(path: Path): Path = {
+    var tablePath = path.getParent;
+    if (tablePath.getName == "_delta_log") {
+      // we do this conditionally as tests in LogStoreSuiet do not
+      // create _delta_log directory
+      tablePath = tablePath.getParent
+    }
+    return tablePath
+  }
```

Comment:
I don't understand why we need this. why don't other LogStores need this to pass unit tests?

Reply:
I don't like this 'hack' too, we should just have `tablePath = path.getParent.getParent` here.

But without that two tests are failing ('detects conflict' and 'listFrom') with following stacktrace:

```
java.nio.file.FileAlreadyExistsException: TransactionLog exists file:/tmp/_delta_log/0
[info]   at io.delta.storage.MemoryLogStore.putExternalEntry(ExternalLogStoreSuite.scala:162)
[info]   at io.delta.storage.BaseExternalLogStore.write(BaseExternalLogStore.scala:266)
```

The reason is these tests create paths this way:

```
    val deltas = Seq(0, 1).map(i => new File(tempDir, i.toString)).map(_.toURI).map(new Path(_))
```

so files are created directly in test's temporary directory and `path.getParent.getParent` goes outside of test's temporary directory (and this directory, common to 3 tests is taken as `tablePath`)

The proper solution for that is to create these files inside subdirectory of temp dir in 'read / write', 'detects conflicts' and 'listFrom' tests and I can fix it this way.

Reply:
> The proper solution for that is to create these files inside subdirectory of temp dir in 'read / write', 'detects conflicts' and 'listFrom' tests and I can fix it this way.

That would be preferable.